orderer/consensus/hlmirbft/chain.go

```diff
@@ -1056,146 +1103,206 @@ func (c *Chain) triggerCatchup(sn *raftpb.Snapshot) {
 func (c *Chain) Apply(batch *msgs.QEntry) error {
 	c.pendingBatches[batch.SeqNo] = batch
-	index := 0  // Review comment change to rpelace append by index.
+	index := 0 // Review comment change to rpelace append by index.
 	seqNumbers := make([]uint64, len(c.pendingBatches))
 	for k := range c.pendingBatches {
 		seqNumbers[index] = k
 		index++
 	}
 	sort.SliceStable(seqNumbers, func(i, j int) bool { return seqNumbers[i] < seqNumbers[j] })
-	for i:=0;i<len(seqNumbers);i++{
-			if c.Node.LastCommittedSeqNo+1 != seqNumbers[i] {
-				break
-			}
-			err := c.processBatch(c.pendingBatches[seqNumbers[i]])
-			if err != nil {
-			    return errors.WithMessage(err, "Batch Processing Error")
-			}
-			delete(c.pendingBatches, seqNumbers[i])
-			c.Node.LastCommittedSeqNo++
+	for i := 0; i < len(seqNumbers); i++ {
+		if c.Node.LastCommittedSeqNo+1 != seqNumbers[i] {
+			break
+		}
+		err := c.processBatch(c.pendingBatches[seqNumbers[i]])
+		if err != nil {
+			return errors.WithMessage(err, "Batch Processing Error")
+		}
+		delete(c.pendingBatches, seqNumbers[i])
+		c.Node.LastCommittedSeqNo++
 	}
 
 	return nil
 }
+
 //FLY2-48 proposed changes
 //	- create Blocks
-func (c *Chain ) CreateBlock(envs []*common.Envelope) *common.Block {
+func (c *Chain) CreateBlock(envs []*common.Envelope) *common.Block {
+
 	bc := &blockCreator{
 		hash:   protoutil.BlockHeaderHash(c.lastBlock.Header),
+		//change
 		number: c.lastBlock.Header.Number,
 		logger: c.logger,
 	}
 	return bc.createNextBlock(envs)
 }
 
+//JIRA FLY2-106 check if pending batch list is empty
+func (c *Chain) isPendingBatchesEmpty() bool {
+	if len(c.pendingBatches) == 0 {
+		return true
+	}
+	return false
+}
+
+//JIRA FLY2-106 sleep till pending batch list is empty
+func (c *Chain) waitForPendingBatchCommits() {
+	isEmpty := c.isPendingBatchesEmpty()
+	for !isEmpty {
+		time.Sleep(400 * time.Millisecond)
+		isEmpty = c.isPendingBatchesEmpty()
+	}
+}
 
 //JIRA FLY2-66 proposed changes:Implemented the Snap Function
 func (c *Chain) Snap(networkConfig *msgs.NetworkState_Config, clientsState []*msgs.NetworkState_Client) ([]byte, []*msgs.Reconfiguration, error) {
-	
-	pr := c.Node.PendingReconfigurations
-	
-	c.Node.PendingReconfigurations = nil
+	var pr []*msgs.Reconfiguration
+	var newNetworkConfig *msgs.NetworkState_Config
+	//JIRA - 106 check reconfiguration length
+	if len(c.Node.PendingReconfigurations) != 0 {
+		pr = c.Node.PendingReconfigurations[0]
+		newNetworkConfig = pr[1].GetNewConfig()
+
+		//JIRA FLY2-106 wait till pending batch list is empty
+		c.waitForPendingBatchCommits()
+		if !reflect.DeepEqual(newNetworkConfig, networkConfig) {
+			//JIRA FLY2-106 get config envelope
+			env, err := c.getOrSetConfigEnv(newNetworkConfig, nil)
+			if err != nil {
+				return nil, nil, errors.WithMessage(err, "Could not retrieve payload")
+			}
+			b := c.CreateBlock([]*common.Envelope{env})
+			c.writeConfigBlock(b)
+			networkConfigHash, _ := c.getMsgHash(newNetworkConfig)
```

Comment:
for completeness wrap this line and the next into a function which removes the pending config envelope - `removeConfigEnv`

Reply:
@harrymknight  issue resolved